sipyco/pc_rpc.py

```diff
@@ -296,126 +371,135 @@ class BestEffortClient:
+        # Python fails to provide a way to cancel threads...
+        self.__conmonitor_terminate = True
 
-    def __send(self, obj):
-        line = pyon.encode(obj) + "\n"
-        self.__socket.sendall(line.encode())
-
-    def __recv(self):
-        buf = self.__socket.recv(4096).decode()
-        while "\n" not in buf:
-            more = self.__socket.recv(4096)
-            if not more:
-                break
-            buf += more.decode()
-        return pyon.decode(buf)
+    def is_closed(self):
+        if not self.__client:
+            return True
 
-    def __do_rpc(self, name, args, kwargs):
-        if self.__conretry_thread is not None:
-            return None
-
-        obj = {"action": "call", "name": name, "args": args, "kwargs": kwargs}
-        try:
-            self.__send(obj)
-            obj = self.__recv()
-        except:
-            logger.warning("connection failed while attempting "
-                           "RPC to %s:%d[%s], re-establishing connection "
-                           "in the background",
-                           self.__host, self.__port, self.__target_name)
-            self.__start_conretry()
-            return None
-        else:
-            if obj["status"] == "ok":
-                return obj["ret"]
-            elif obj["status"] == "failed":
-                raise_packed_exc(obj["exception"])
-            else:
-                raise ValueError
+        with self.__client_lock:
+            return self.__client.is_closed()
 
     def __getattr__(self, name):
-        if name not in self.__valid_methods:
-            raise AttributeError
-        def proxy(*args, **kwargs):
-            return self.__do_rpc(name, args, kwargs)
-        return proxy
-
-    def get_selected_target(self):
-        raise NotImplementedError
+        """
+        If the client is connected, pass this call onto the client. If not,
+        return a method that returns None if the call was valid or if we can't
+        tell. Raise AttributeError otherwise.
+        """
 
-    def get_local_host(self):
-        raise NotImplementedError
+        # If the client exists, pass the call to the client
+        if self.__client:
+            client_method = getattr(self.__client, name)
+
+            def tryer(*args, **kwargs):
+                with self.__client_lock:
+                    try:
+                        return client_method(*args, **kwargs)
+                    except Exception:
+                        self.__mark_invalid()
+                        return None
+            return tryer
+        elif (
+            self.__valid_methods
+            and name not in self.__valid_methods
+            and name not in dir(Client)
+        ):
+            raise AttributeError("{} not found in cached list".format(name))
+        else:
+            def none_returner(*args, **kwargs):
+                return None
+            return none_returner
```

Comment:
This can make errors unnecessarily persistent, no?
```python
x = client.foo
while True:
  x()
```

Reply:
Yep, you're right. I've pushed some changes to resolve that by checking for the connection on each call.